py/server/deephaven/table_listener.py

```diff
@@ -393,3 +397,205 @@ def listen(t: Table, listener: Union[Callable, TableListener], description: str
+                None.
+
+                Dependencies ensure that the listener can safely access the dependent tables during its execution. This
+                mainly includes reading the data from the tables. While performing operations on the dependent tables in
+                the listener is safe, it is not recommended because reading or operating on the result tables of those
+                operations may not be safe. It is best to perform the operations on the dependent tables beforehand,
+                and then add the result tables as dependencies to the listener so that they can be safely read in it.
+
+        Raises:
+            DHError
+        """
+        if len(tables) < 2:
+            raise DHError(message="A merged listener must listen to at least 2 refreshing tables.")
+
+        self.tables = tables
+        self.listener_recorders = [_ListenerRecorder(t) for t in tables]
+
+        self.dependencies = dependencies
+
+        if isinstance(listener, MergedListener):
+            self.listener = listener.on_update
+        else:
+            self.listener = listener
+        n_params = len(signature(self.listener).parameters)
+        if n_params != 2:
+            raise ValueError("merged listener function must have 2 parameters (updates, is_replay).")
+
+
+        try:
+            self.merged_listener_adapter = _JPythonMergedListenerAdapter.create(
+                        to_sequence(self.listener_recorders),
+                        to_sequence(self.dependencies),
+                        description,
+                        self)
+            self.started = False
+        except Exception as e:
+            raise DHError(e, "failed to create a merged listener adapter.") from e
+
+    def _process(self, replay_updates: Optional[jpy.JType] = None) -> None:
+        """Process the table updates from the listener recorders.
+
+        Args:
+            replay_updates (Optional[jpy.JType]): the replay updates, only provided during replay.
+        """
+        if replay_updates:
+            self.listener({t: TableUpdate(t, tu) for t, tu in zip(self.tables, j_list_to_list(replay_updates))}, True)
+        else:
+            self.listener({lr.table: lr.table_update() for lr in self.listener_recorders}, False)
+
+    def start(self, do_replay: bool = False, replay_lock: Literal["shared", "exclusive"] = "shared") -> None:
+        """Start the listener by registering it with the tables and listening for updates.
+
+        Args:
+            do_replay (bool): whether to replay the initial snapshots of the tables, default is False
+            replay_lock (str): the lock type used during replay, default is 'shared', can also be 'exclusive'.
+
+        Raises:
+            DHError
+        """
+        if self.started:
+            raise RuntimeError("Attempting to start an already started merged listener..")
+
+        try:
+            def _start():
```

Comment:
I find this locking pattern quite confusing. Just use auto locking context over the whole thing. Users will most likely already have the exclusive lock. We don't need an argument to tell us how to lock internally; we have explicit constructs for controlling locking.